core/include/gap/core/stream.hpp

```diff
@@ -0,0 +1,61 @@
+// Copyright (c) 2022-present, Trail of Bits, Inc.
+
+#pragma once
+
+#ifdef GAP_ENABLE_COROUTINES
+
+    #include "gap/core/concepts.hpp"
+    #include "gap/core/generator.hpp"
+
+    #include <optional>
+
+namespace gap
+{
+    namespace stream
+    {
+        template< typename T >
+        requires iterable< T >
+        auto stream(T&& cont) -> gap::generator< decltype(*cont.begin()) > {
+            for (auto&& elem : cont) {
+                co_yield std::forward< decltype(elem) >(elem);
+            }
+        }
+
+        template< typename T >
+        requires std::convertible_to< T, bool > && dereferenceable< T >
+        auto stream(T&& cont) -> gap::generator< decltype(*cont) > {
+            if (cont) {
+                co_yield std::forward< decltype(*cont) >(*cont);
+            }
+        }
+
+        template< typename TFunc, typename TInput >
+        auto map(TFunc&& f, gap::generator< TInput > source)
+            -> gap::generator< decltype(f(*source.begin())) > {
+            for (auto&& elem : source) {
+                co_yield f(std::forward< decltype(elem) >(elem));
+            }
+        }
+
+        template< typename TPred, typename TElem >
+        gap::generator< TElem > filter(TPred&& pred, gap::generator< TElem > source) {
+            for (auto&& elem : source) {
+                if (pred(std::forward< decltype(elem) >(elem))) {
```

Comment:
forward here creates a dangling reference, you never want to move into predicate